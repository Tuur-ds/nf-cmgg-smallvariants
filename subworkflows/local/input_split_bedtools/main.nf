//
// Split the BED files and return a channel ready to use for scatter/gather flows
//

include { BEDTOOLS_SPLIT } from '../../../modules/nf-core/bedtools/split/main'

workflow INPUT_SPLIT_BEDTOOLS {
    take:
        ch_beds     // channel: [mandatory] [ val(meta), path(bed) ] => bed files
        ch_inputs   // channel: [mandatory] [ val(meta), path(input), path(input_index), path(extra_file) ] => input files

    main:

    ch_versions = Channel.empty()

    BEDTOOLS_SPLIT(
        ch_beds.map { it + [params.scatter_count]}
    )
    ch_versions = ch_versions.mix(BEDTOOLS_SPLIT.out.versions.first())

    ch_inputs
        .join(BEDTOOLS_SPLIT.out.beds, failOnDuplicate: true, failOnMismatch: true)
        .map { meta, input, input_index, extra_file, beds ->
            // Determine the amount of BED files per sample
            bed_is_list = beds instanceof ArrayList
            new_meta = meta + [split_count: bed_is_list ? beds.size() : 1]
            [ new_meta, input, input_index, extra_file, bed_is_list ? beds : [beds] ]
        }
        .transpose(by:3) // Create one channel entry for each BED file per sample
        .map { meta, input, input_index, extra_file, bed ->
            // Set the base name of the BED file as the ID (this will look like sample_id.xxxx, where xxxx are numbers)
            new_meta = meta + [id:bed.baseName]
            [ new_meta, input, input_index, extra_file, bed ]
        }
        .set { ch_split_output }

    emit:
        split = ch_split_output // channel: [ val(meta), path(input), path(input_index), path(extra_file), path(bed) ]
        versions = ch_versions  // channel: [ versions.yml ]
}