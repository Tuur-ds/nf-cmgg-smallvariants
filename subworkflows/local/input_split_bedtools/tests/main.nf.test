nextflow_workflow {

    name "Test Workflow INPUT_SPLIT_BEDTOOLS"
    script "../main.nf"
    workflow "INPUT_SPLIT_BEDTOOLS"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "input_split_bedtools"

    test("input_split_bedtools - default - crams") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [id:"NA24143", sample:"NA24143", family:"Ashkenazim", family_samples:"NA24143"],
                    file(params.bed, checkIfExists:true),
                    3
                ])
                input[1] = Channel.of([
                    [id:"NA24143", sample:"NA24143", family:"Ashkenazim", family_samples:"NA24143"],
                    file(params.cram1, checkIfExists:true),
                    file(params.crai1, checkIfExists:true),
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.split.collect { it.collect { it instanceof Map || it.endsWith(".bed") ? it : file(it).name } }
                ).match("default - crams") }
            )
        }

    }
}
