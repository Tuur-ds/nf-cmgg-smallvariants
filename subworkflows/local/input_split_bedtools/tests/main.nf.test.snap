{
    "default - crams": {
        "content": [
            [
                [
                    {
                        "id": "NA24143.00001",
                        "sample": "NA24143",
                        "family": "Ashkenazim",
                        "family_samples": "NA24143",
                        "split_count": 3
                    },
                    "NA24143.cram",
                    "NA24143.cram.crai",
                    "NA24143.00001.bed:md5,ffd6f70b7fa85ffb46fb66a4cf9afb70"
                ],
                [
                    {
                        "id": "NA24143.00002",
                        "sample": "NA24143",
                        "family": "Ashkenazim",
                        "family_samples": "NA24143",
                        "split_count": 3
                    },
                    "NA24143.cram",
                    "NA24143.cram.crai",
                    "NA24143.00002.bed:md5,628a9845316c9bdf2e8c11c04bdd0f06"
                ],
                [
                    {
                        "id": "NA24143.00003",
                        "sample": "NA24143",
                        "family": "Ashkenazim",
                        "family_samples": "NA24143",
                        "split_count": 3
                    },
                    "NA24143.cram",
                    "NA24143.cram.crai",
                    "NA24143.00003.bed:md5,6eabb4f395fe99ff241aaa3391978b06"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-05T17:39:07.719231196"
    }
}